exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java

```diff
@@ -0,0 +1,271 @@
+              .setCredentialsProvider(FixedCredentialsProvider.create(credentials))
+              .build());
+    }
+    return new StackdriverExporter(
+        projectId, metricServiceClient, exportInterval, Stats.getViewManager());
+  }
+
+  /**
+   * Register a {@link View} against this exporter, and upload it as a {@link MetricDescriptor} to
+   * StackDriver.
+   *
+   * @param view the {@code View} to be registered.
+   */
+  public void registerView(View view) {
+    workerThread.registerView(view);
+    // TODO(songya): don't need to create MetricDescriptor for RpcViewConstants once we defined
+    // canonical metrics. Registration is required only for custom view definitions. Canonical views
+    // should be pre-registered.
+    MetricDescriptor metricDescriptor = createMetricDescriptor(view, projectId);
+    if (metricDescriptor != null) {
+      metricServiceClient.createMetricDescriptor(
+          CreateMetricDescriptorRequest
+              .newBuilder()
+              .setMetricDescriptor(metricDescriptor)
+              .build());
+    }
+  }
+
+  // Worker thread that polls ViewData from Stats library and batch export to StackDriver.
+  private static final class WorkerThread extends Thread {
+
+    private final long scheduleDelayMillis;
+    private final String projectId;
+    private final MetricServiceClient metricServiceClient;
+    private final ViewManager viewManager;
+
+    private static final Object monitor = new Object();
+
+    @GuardedBy("monitor")
+    private final Map<Name, View> registeredViews = new HashMap<Name, View>();
+
+    private WorkerThread(
+        String projectId,
+        MetricServiceClient metricServiceClient,
+        Duration exportInterval,
+        ViewManager viewManager) {
+      this.scheduleDelayMillis = toMillis(exportInterval);
+      this.projectId = projectId;
+      this.metricServiceClient = metricServiceClient;
+      this.viewManager = viewManager;
+      setDaemon(true);
+      setName("ExportWorkerThread");
+    }
+
+    private void registerView(View view) {
+      synchronized (monitor) {
+        View existing = registeredViews.get(view.getName());
+        if (existing != null) {
+          if (existing.equals(view)) {
+            // Ignore views that are already registered.
+            return;
+          } else {
+            throw new IllegalArgumentException(
+                "A different view with the same name is already registered: " + existing);
+          }
+        }
+        registeredViews.put(view.getName(), view);
+      }
+    }
+
+    // Polls ViewData from Stats library for all registered views, and upload them as TimeSeries to
+    // StackDriver.
+    private void export() {
+      CreateTimeSeriesRequest.Builder builder = CreateTimeSeriesRequest.newBuilder();
+      synchronized (monitor) {
+        for (View.Name view : registeredViews.keySet()) {
+          builder.addAllTimeSeries(createTimeSeriesList(viewManager.getView(view), projectId));
```

Comment:
To avoid holding the lock for too long better to just extract the ViewData, add to a list then outside the lock create the proto.

Reply:
Updated